# R script
# 1. Evaluates read depth across genomic intervals

```r
#input files: Samples metadata and CNVnator read depth files estimated from BAM files at each CNV
#output files: list of CNVs with population information, biallelic CNVs, and figures displaying estimated read depth

library(stringr)
library(plyr)
library(ggplot2)
library(reshape)
library(dplyr)
library(fpc)

########################################################################################################################
###############################################################################################################
samples<-read.table("Samples96.out",col.names=c("sample","sex","pop"))
sampleorder<-samples$sample[order(samples$pop,samples$sex)]
samples$sample<-factor(samples$sample,levels=sampleorder)

##pairwise difference function
pairwisediff<-function(x){
  rank=sort(x)
  diff=max(diff(as.numeric(rank)))
  return(diff)
}

################################################################################################################################################################################
################ Read through CNVnator Read Depth in BED format from BAM files for each individual at each deletion or duplication call
##filters: if(maX<1.5 | miN > 2.5 | mdiff<1 | diff<dthreshold)

## Loop through each CNV
if (!file.exists("Figures")){dir.create("Figures") }# create a new sub directory for figures
for(CNVfile in c("All96.genotypes.deletions.bed","All96.genotypes.duplications.bed")){

  cnvs<-read.delim(CNVfile,header=T,sep="\t",comment.char="")
  list=c(4:ncol(cnvs))# this is the column number of all lines

  ### Look for ratio of minimum:non-minimums =< 0.5, assuming that others have at least 
  ### twice as much coverage on average (also assumes 1/few have deletion)
  
  dthreshold=0.45 ##this is the spread - there should be at least 1 gap between two individuals that spans this depth
  rm(newtable)
  len<-length(list)
  
  ### Takes in all CNV genotyping (copy-number) results and combines them
  for (row in 1:nrow(cnvs)){
    miN<-min(cnvs[row,list])
    maX=max(cnvs[row,list])
    suM=sum(cnvs[row,list])
    avG=suM/len
    mdiff=maX-miN
    
    ## look for gap between coverage, filter based on threshold. This is to prevent noise from overwhelming calls
    rank=sort(cnvs[row,list])
    diff=0
    #get pairwise differences
    for(i in 2:ncol(rank)){
      if((rank[i]-rank[i-1])>diff){diff=(rank[i]-rank[i-1])}
    }
    ###exclude really low coverage or if max - min is less than 1
    if(maX<1.5 | miN > 2.5 | mdiff<1 | diff<dthreshold){
      next
    }
   
    ### Round to Integer
    Int<-round(cnvs[row,list])
    miN<-min(Int)
    maX=max(Int)
    suM=sum(Int)
    avG=suM/len
    mdiff=maX-miN
    num0<-sum(cnvs[row,list]==0) ##how many have 0
      
    ###If max is 1, then 0 are deletions
    
    ##0 and 1 are deletions, whereas 3 and higher are duplications
      ###Add to vectors which individuals are deletions or duplications
      ndels<-vector()
      ndups<-vector()
      nde=0
      ndu=0
      for(ind in 1:len){
        nind<-colnames(Int[ind])
        if(Int[,ind]<2){
          ndels<-c(ndels,nind)
          nde=nde+1
        }
        if(Int[,ind]>2){
          ndups<-c(ndups,nind)
          ndu=ndu+1
        }
      }
    
    if(exists("ndels") | exists("ndups")){
      ###Make row, with list of deletions and duplications
      ccc<-cnvs[row,]
      ccc$ndel<-nde
      ccc$ndup<-ndu
      ccc$deletions<-paste(ndels,collapse=",")
      ccc$duplications<-paste(ndups,collapse=",")
      
      ##Cluster using kmeansruns from the package fpc 
      tt<-t(cnvs[row,order(samples$sample)+3]) ##resort by population and sex
      
      ##To pick the best value of k, weâ€™ll use kmeansruns function from fpc package. 
      ##This function is enabled with two distance metrics: Average silhouette width and Calinski-Harabas
      clus3<-as.data.frame(cbind(X1=row.names(tt),value=tt[,1],pop=samples$pop[order(samples$sample)]))
      clus3$value<-as.numeric(clus3$value)
      clus3<-merge(clus3,samples,by.x=c("X1","pop"),by.y=c("sample","pop"),sort=F)
        
      if(96-sum(cnvs[row,list]==0)>10){
        tunekw <- kmeansruns(tt,krange = 2:10,criterion = "asw")
        tunekw$bestk
        
        ##get clusters
        clusters<-cutree(hclust(dist(t(cnvs[row,order(samples$sample)+3])),method="ward.D2"), k = tunekw$bestk)
        clus<-as.data.frame(cbind(tt,clusters))
        colnames(clus)<-c("coverage","cluster")
        
        clus2<-cbind(melt(as.matrix(clus[1])),clusters,pop=samples$pop[order(samples$sample)])
        clus2$X1<-factor(clus2$X1,levels=clus2$X1)
        clus2<-merge(clus2,samples,by.x=c("X1","pop"),by.y=c("sample","pop"),sort=F)
        
        ##output figures for visual inspection and setting manual thresholds
        ggplot(clus2,aes(X1,value))+
          geom_point(colour=clusters,size=2,aes(shape = factor(sex))) +
          facet_grid(~pop,scale="free") +
          ylab("read depth") + xlab("sample") +
          ylim(0,max(clus2$value)) +
          theme_light(base_size = 16) +
          ggtitle(paste(cnvs[row,1],cnvs[row,2],cnvs[row,3])) +
          theme(axis.text.x = element_text(angle = 90, size = 5),legend.position = "",
                panel.spacing.x=unit(0.2,"lines"))
        ggsave(paste("Figures/Depth",cnvs[row,1],cnvs[row,2],cnvs[row,3],"clusters.pdf",sep="."),width=6)
      } 
      
      ##pre-made manual clusters as a guide, based on strict thresholds, DEL=1.5 and DUP=2.5
      clus3$clusters[clus3$value>4.5]<-6
      clus3$clusters[clus3$value<=4.5]<-5
      clus3$clusters[clus3$value<=3.5]<-4
      clus3$clusters[clus3$value<=2.5]<-3
      clus3$clusters[clus3$value<=1.5]<-2
      clus3$clusters[clus3$value<0.5]<-1
      clus3$clusters<-as.factor(clus3$clusters)
      clusters<-as.factor(clus3$clusters)
      clus3$X1<-factor(clus3$X1,levels=clus3$X1)
      
      ggplot(clus3,aes(X1,value))+
        geom_point(color=clusters,size=2,aes(shape = factor(sex))) +
        facet_grid(~pop,scale="free") +
        ylab("read depth") + xlab("sample") +
        ylim(0,max(clus3$value)) +
        theme_light(base_size = 16) +
        ggtitle(paste(cnvs[row,1],cnvs[row,2],cnvs[row,3])) +
        theme(axis.text.x = element_text(angle = 90, size = 5),legend.position = "",
              panel.spacing.x=unit(0.2,"lines"))
      ggsave(paste("Figures/Depth",cnvs[row,1],cnvs[row,2],cnvs[row,3],"clusters.manual.pdf",sep="."),width=6)
      
      ###add to table info, then also add info from clusters
      ccc$ncnv<-96-length(clus3$X1[clus3$clusters==2])
      ccc$del0<-paste(clus3$X1[clus3$clusters==0],collapse=";")
      ccc$del1<-paste(clus3$X1[clus3$clusters==1],collapse=";")
      ccc$dup3<-paste(clus3$X1[clus3$clusters==3],collapse=";")
      ccc$dup4<-paste(clus3$X1[clus3$clusters==4],collapse=";")
      ccc$dup5<-paste(clus3$X1[clus3$clusters==5],collapse=";")
      
      if(exists("newtable")){
        newtable<-rbind(newtable,ccc)
      }else{
        newtable<-ccc
      }
    }
  }
  
  ################################################################################################
  ##### Add Population Del / Dup numbers
  for(deldup in c("del","dup")){
    for(po in sort(unique(samples$pop))){
      ##how many of each pop in table are deletions
      if(deldup=="del"){
        newtable[,paste0(po,"del")]<-as.vector(colSums(apply(newtable[,names(newtable) %in% samples$sample[samples$pop==po]], 1, function(x) x<1.5)))
      }
      ##how many of each pop in table are duplications
      if(deldup=="dup"){
        newtable[,paste0(po,"dup")]<-as.vector(colSums(apply(newtable[,names(newtable) %in% samples$sample[samples$pop==po]], 1, function(x) x>2.5)))
      }
    }
  }
  write.table(file=paste(CNVfile,"final.table.out",sep="."),newtable,row.names=F,quote=F,sep="\t")

  ################################################################################################
  ################################ PCA  ################################################

  popsorder2<-as.data.frame(samples$pop)
  row.names(popsorder2)<-colnames(newtable[,4:99])
  colnames(popsorder2)<-"pop"
  
  #binary
  cnvpca<-round(newtable[,4:99])
  cnvpca_binary<-cnvpca
  cnvpca_binary[cnvpca_binary<2]<-0
  cnvpca_binary[cnvpca_binary==2]<-1
  cnvpca_binary[cnvpca_binary>2]<-2
  
  #bi-allelic ... only 3 genotypes (CN)
  cnvpca_biallelic<-cnvpca[apply(cnvpca, 1, function(x) n_distinct(x))<4,]
  rounded<-newtable #round read depth to integers
  rounded[,4:99]<-cnvpca
  biallelic<-rounded[apply(rounded[,4:99], 1, function(x) n_distinct(x))<4,]
  write.table(file=paste(CNVfile,"final.table.biallelic.out",sep="."),biallelic,row.names=F,quote=F,sep="\t")

  ################################################################################################
  
}##end for del and dup
```
